commit d1ed37c3a772b2609faaf78717816faf41827629
Author: P33M <p33m@github.com>
Date:   Wed May 1 17:04:32 2019 +0100

    smsc95xx: dynamically fix up TX buffer alignment with padding bytes
    
    dwc_otg requires a 32-bit aligned buffer start address, otherwise
    expensive bounce buffers are used. The LAN951x hardware can skip up to
    3 bytes between the TX header and the start of frame data, which can
    be used to force alignment of the URB passed to dwc_otg.
    
    As found in https://github.com/raspberrypi/linux/issues/2924

diff --git a/drivers/net/usb/smsc95xx.c b/drivers/net/usb/smsc95xx.c
index a5bbec75f22d..c087b6d75753 100644
--- a/drivers/net/usb/smsc95xx.c
+++ b/drivers/net/usb/smsc95xx.c
@@ -2082,7 +2082,9 @@ static struct sk_buff *smsc95xx_tx_fixup(struct usbnet *dev,
 					 struct sk_buff *skb, gfp_t flags)
 {
 	bool csum = skb->ip_summed == CHECKSUM_PARTIAL;
-	int overhead = csum ? SMSC95XX_TX_OVERHEAD_CSUM : SMSC95XX_TX_OVERHEAD;
+	unsigned int align_bytes = -((uintptr_t)skb->data) & 0x3;
+	int overhead = csum ? SMSC95XX_TX_OVERHEAD_CSUM + align_bytes
+				: SMSC95XX_TX_OVERHEAD + align_bytes;
 	u32 tx_cmd_a, tx_cmd_b;
 
 	/* We do not advertise SG, so skbs should be already linearized */
@@ -2116,16 +2118,16 @@ static struct sk_buff *smsc95xx_tx_fixup(struct usbnet *dev,
 		}
 	}
 
-	skb_push(skb, 4);
-	tx_cmd_b = (u32)(skb->len - 4);
+	skb_push(skb, 4 + align_bytes);
+	tx_cmd_b = (u32)(skb->len - 4 - align_bytes);
 	if (csum)
 		tx_cmd_b |= TX_CMD_B_CSUM_ENABLE;
 	cpu_to_le32s(&tx_cmd_b);
 	memcpy(skb->data, &tx_cmd_b, 4);
 
 	skb_push(skb, 4);
-	tx_cmd_a = (u32)(skb->len - 8) | TX_CMD_A_FIRST_SEG_ |
-		TX_CMD_A_LAST_SEG_;
+	tx_cmd_a = (u32)(skb->len - 8 - align_bytes) | TX_CMD_A_FIRST_SEG_ |
+		(align_bytes << 16) | TX_CMD_A_LAST_SEG_;
 	cpu_to_le32s(&tx_cmd_a);
 	memcpy(skb->data, &tx_cmd_a, 4);
 
diff --git a/drivers/net/usb/smsc95xx.h b/drivers/net/usb/smsc95xx.h
index cfc704f3a460..55f9f03e1a0c 100644
--- a/drivers/net/usb/smsc95xx.h
+++ b/drivers/net/usb/smsc95xx.h
@@ -21,7 +21,7 @@
 #define _SMSC95XX_H
 
 /* Tx command words */
-#define TX_CMD_A_DATA_OFFSET_	(0x001F0000)	/* Data Start Offset */
+#define TX_CMD_A_DATA_OFFSET_	(0x00030000)	/* Data Start Offset */
 #define TX_CMD_A_FIRST_SEG_	(0x00002000)	/* First Segment */
 #define TX_CMD_A_LAST_SEG_	(0x00001000)	/* Last Segment */
 #define TX_CMD_A_BUF_SIZE_	(0x000007FF)	/* Buffer Size */
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1485_d1ed37c3a772b2609faaf78717816faf41827629.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1485_d1ed37c3a772b2609faaf78717816faf41827629.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
